src/main/java/com/uber/cadence/internal/sync/WorkflowThreadImpl.java

```diff
@@ -143,6 +148,21 @@ public void setName(String name) {
         thread.setName(name);
       }
     }
+
+    public void propagateContext() {
+      // Loop through the context propagators and propagate
+      Map<String, Object> propagatedContexts = decisionContext.getPropagatedContexts();
```

Comment:
I might be confused. But it looks to me  that the context are coming not from a parent thread but from a single decision context shared by all threads. Am I missing something?

Reply:
This might be a long answer but I hope it explains what my overall design was for this feature. 

This is my understanding (very high-level) of the flow of a workflow/activity execution: 

```
1 Workflow client ->
2    [Cadence workflow stub -> Cadence Server -> Workflow worker] ->
3       Workflow Implementation -> 
4            [Cadence activity stub -> Cadence Server -> Activity worker] ->
5                Activity Implementation
```
Where lines 1, 3, and 5 are code written by workflow/activity implementers, and lines 3 & 5 are Cadence infrastructure code.  

My thought was that context should flow only from line 1 to 3 to 5, and that the Cadence infrastructure/stubs/server should be "invisible" to the context propagation.  So when a workflow worker receives a `WorkflowExecutionStartedEvent`, I want to propagate the context into the thread that is executing the Workflow Implementation, not necessarily any of the intermediate threads spawned by the Cadence infrastructure. 

So I guess a) does that make sense, b) is that what you want, and c) am I missing something in that implementation?

Reply:
My idea (may be flawed) is that any thread should be able to access and modify the context. For example it can create a new tracing span and all the code that executes in that span including child threads should inherit from that context. So from the context propagation point of view the context should come from the parent thread, not from the workflowExecutionStartedEvent which injects context only into the main workflow thread.

Reply:
I thought of doing that briefly, but that seemed to imply that everywhere Cadence started a new thread, we'd have to pass around the ContextPropagators and the context data.  That seemed like a lot of parameters moving around everywhere. 

I did have a thought this morning - I suppose I could use a ThreadLocal to hold the list of ContextPropagators and the context data.  But that would still require every Runnable object to take extra parameters for the context information.  

If you'd prefer that approach I can work on it.

Reply:
Ahh - one thing I hadn't seen until now was the `Workflow.newPromise` and similar methods that allowed the Workflow to spawn off threads.  I'll do some more updates that take those into account.

Reply:
I don't think newPromise spawns threads. Only Async.... methods do. Use WorkflowThreadLocal instead of ThreadLocal.

Reply:
Ok - I have it working now where any child threads will inherit the context from the parent threads.  Let me know what you think